### ISSUE TITLE
Add ignore: ["custom-properties"] to length-zero-no-unit

### ISSUE THREAD

**Issue by jethrolarson** [stylelint/stylelint#2586]
A bug where using stylefmt with stylelint causes calc expressions to break

> Which rule, if any, is this issue related to?

length-zero-no-unit 

> What CSS is needed to reproduce this issue?

```css
.foo {
  --x: 0px;
  left: calc(300px - var(--x));
}
```

> What stylelint configuration is needed to reproduce this issue?

```yaml
extends: stylelint-config-standard
```

> Which version of stylelint are you using?

7.10.1

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylefmt -r 'app/**/*.scss' 'app/**/*.css'`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)? 
no

> What did you expect to happen?

Fixing the warning to be safe since it's in the recommended list

> What actually happened (e.g. what warnings or errors you are getting)?

Styles get fixed to
```css
.foo {
  --x: 0;
  left: calc(300px - var(--x));
}
```
calc sees different units on left and right hand side of the operator and returns a 0.

I think that this rule should be removed from the recommended list since it is not safe to fix.

I'm filing this with stylefmt as well.

**Comment by hudochenkov** [comment]
Thank you for reporting. This use case will be helpful for us, if we're going to implement fix functionality for this rule.

But currently stylelint doesn't have fixing functionality. Please, report this problem to stylefmt repository.

**Comment by hudochenkov** [comment]
On the other hand, stylelint will show violation if `length-zero-no-unit` is enabled:

```css
.foo {
  --x: 0px;
  left: calc(300px - var(--x));
}
```

@stylelint/core do you think we should ignore custom properties in this rule? Or it's rare case and user could disable a rule for required line?

**Comment by evilebottnawi** [comment]
@hudochenkov from spec:
```
For example, the following code incorrectly attempts to use a variable as a property name:
.foo {
  --side: margin-top;
  var(--side): 20px;
}
This is not equivalent to setting margin-top: 20px;. Instead, the second declaration is simply thrown away as a syntax error for having an invalid property name.

Similarly, you can’t build up a single token where part of it is provided by a variable:

.foo {
  --gap: 20;
  margin-top: var(--gap)px;
}
Again, this is not equivalent to setting margin-top: 20px; (a length). Instead, it’s equivalent to margin-top: 20 px; (a number followed by an ident), which is simply an invalid value for the margin-top property. Note, though, that calc() can be used to validly achieve the same thing, like so:

.foo {
  --gap: 20;
  margin-top: calc(var(--gap) * 1px);
}
```
Seems it is valid and stylelint do it right.

**Comment by jeddy3** [comment]
The following looks like valid code _and a valid use case_ to me:

```css
.foo {
  --x: 0px;
  left: calc(300px - var(--x));
}
```

I can see myself using that more than I use the following:

```css
.foo {
  --x: 0;
  left: calc(300px - (var(--x) * 1px);
}
```

I think providing an option, `ignore: ["custom-properties"]`, to make the rule more lenient would be in line with our approach elsewhere in stylelint.

@jethrolarson Would you like to [contribute this option](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-an-option-to-an-existing-rule)? It should be pretty trivial and a great way to help the project. It should just be the case of returning early [here](https://github.com/stylelint/stylelint/blob/master/lib/rules/length-zero-no-unit/index.js#L28) if the `ignore: ["custom-properties"]` option is set. Look [here](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-type-no-unknown/index.js#L31) and [here](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-type-no-unknown/index.js#L63-L68) for examples of how options work. Also, [here's](https://github.com/stylelint/stylelint/blob/master/lib/rules/custom-property-no-outside-root/index.js#L35-L37) an example of using the `isCustomProperty` util. Lastly, here are example [docs](https://github.com/stylelint/stylelint/tree/master/lib/rules/selector-type-no-unknown#optional-secondary-options) and [test](https://github.com/stylelint/stylelint/blob/master/lib/rules/selector-type-no-unknown/__tests__/index.js#L260-L283).

**Comment by jitendravyas** [comment]
Any conclusion on this?

**Comment by hudochenkov** [comment]
@jitendravyas add [disable comments](https://stylelint.io/user-guide/configuration/#turning-rules-off-from-within-your-css) to disable rule on particular line. Or consider contributing a new option, as per [comment above](https://github.com/stylelint/stylelint/issues/2586#issuecomment-304629118).

**Comment by CAYdenberg** [comment]
@stylelint/core Should this issue be closed?

**Comment by jeddy3** [comment]
@CAYdenberg Thanks for checking up on these. Yep, should be closed.
